data_diff/dbt.py

```diff
@@ -366,22 +371,56 @@ def _get_connection_creds(self) -> Tuple[Dict[str, str], str]:
 
         return credentials, conn_type
 
+    @staticmethod
+    def _get_snowflake_private_key(credentials):
+        """Get Snowflake private key by path, from a Base64 encoded DER bytestring or None."""
+        if credentials.get("private_key") and credentials.get("private_key_path"):
+            raise Exception("Cannot specify both `private_key`  and `private_key_path`")
+
+        if credentials.get("private_key_passphrase"):
+            encoded_passphrase = credentials.get("private_key_passphrase").encode()
+        else:
+            encoded_passphrase = None
+
+        if credentials.get("private_key"):
+            p_key = serialization.load_der_private_key(
+                base64.b64decode(credentials.get("private_key")),
+                password=encoded_passphrase,
+                backend=default_backend(),
+            )
+        elif credentials.get("private_key_path"):
+            with open(credentials.get("private_key_path"), "rb") as key:
+                p_key = serialization.load_pem_private_key(
+                    key.read(), password=encoded_passphrase, backend=default_backend()
+                )
+        else:
+            return None
+
+        return p_key.private_bytes(
+            encoding=serialization.Encoding.DER,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+
     def set_connection(self):
         credentials, conn_type = self._get_connection_creds()
 
         if conn_type == "snowflake":
-            if credentials.get("password") is None or credentials.get("private_key_path") is not None:
-                raise Exception("Only password authentication is currently supported for Snowflake.")
+            if credentials.get("authenticator") is not None:
+                raise Exception("Federated authentication is not currently supported for Snowflake.")
             conn_info = {
                 "driver": conn_type,
                 "user": credentials.get("user"),
-                "password": credentials.get("password"),
                 "account": credentials.get("account"),
                 "database": credentials.get("database"),
                 "warehouse": credentials.get("warehouse"),
                 "role": credentials.get("role"),
                 "schema": credentials.get("schema"),
             }
+            if credentials.get("password") is not None:
+                conn_info["password"] = credentials.get("password")
+            else:
+                conn_info["private_key"] = self._get_snowflake_private_key(credentials)
```

Comment:
Thank you for the callout - I was not aware of this code!

To be clear, are you asking that I follow the pattern outlined there for error-handling? Happy to follow that example!

Reply:
ah, so remove the private key restriction that is imposed in `dbt.py`. 

It seems by the way the code flows that a wider refactor would be required - it attempts to connect with the provided credentials before ever getting to the downstream code

Our company is trying to evaluate this product; we are unable to do so while data-diff does not support key pair auth for dbt

Would it be possible to contribute this as is? I'm happy to provide further contribution in terms of DRYing out the code in the future

Reply:
Are you saying it's creating a snowflake connection somewhere other than here? https://github.com/datafold/sqeleton/blob/6c532733e1a3efe329b573691d3d9d18dcf4f410/sqeleton/databases/snowflake.py#L182

I may not be following, I can take a closer look later today

Reply:
I am saying that, with the `dbt.py` module, the connection is created before it hits the [this file/repo](https://github.com/datafold/sqeleton/blob/6c532733e1a3efe329b573691d3d9d18dcf4f410/sqeleton/databases/snowflake.py#L156). To exemplify, I have removed any restrictions on providing authentication (see picture)
<img width="777" alt="Screenshot 2023-03-22 at 11 47 46 AM" src="https://user-images.githubusercontent.com/16657925/226993157-616b92fd-39a2-471c-9f1b-db55feb22b7b.png">

If i now run `poetry run python3 data_diff --dbt` while not providing a password but still providing a valid key pair auth, I receive:
```bash
[10:59:51] ERROR - 251006: Password is empty
```

I don't believe we are ever reaching the code you linked as we, seemingly, only run [this code for the connection](https://github.com/datafold/data-diff/blob/master/data_diff/dbt.py#L84-L85)

Reply:
I'm not an expert in this but it appears that in the main function, [dbt has its own branch of logic](https://github.com/datafold/data-diff/blob/master/data_diff/__main__.py#L261-L269) which calls the DBT parser's connector as opposed to sqeleton's connector.  I don't think the code you linked is called in the case of dbt specifically

Reply:
Hey Jake, it definitely is, the entry point for the connection creation starts here. Eventually making its way to sqeleton https://github.com/datafold/data-diff/blob/d21c14014d037ff99ebf21ea0a39fdbc3068d0f2/data_diff/dbt.py#L153

I can take a closer look at this later today, and implement the changes I'm describing if that helps.

Reply:
yea, I was about to recorrect myself. Following the inheritance as an outsider here is non-trivial. I can keep looking but if you get to it that would be great too

Reply:
were you planning on passing the value thru as `key` in this module or making the change in sqeleton?

Reply:
@dlawin is the plan to refactor the sqeleton code while removing exceptions related to the presence of a priv key here? as is, sqeleton does not allow you to specify a keyphrase for a priv key and seems like it could benefit from at least the `_get_snowflake_private_key` function